Detectors/FIT/FDD/simulation/src/Digitizer.cxx

```diff
@@ -59,38 +55,85 @@ void Digitizer::process(const std::vector<o2::fdd::Hit>* hits, o2::fdd::Digit* d
-      Int_t const lastBin = TMath::Min((Int_t)mNBins - 1, (Int_t)((tPhE + 2. * parameters.mPMTransitTime) / mBinSize));
-      //LOG(INFO) << "firstBin = "<<firstBin<<" lastbin "<<lastBin;
-      Float_t const tempT = mBinSize * (0.5f + firstBin) - tPhE;
-      Float_t* p = analogSignal.data() + firstBin;
-      long iStart = std::lround((tempT + 2.0f * parameters.mPMTransitTime) / mBinSize);
-      float const offset = tempT + 2.0f * parameters.mPMTransitTime - Float_t(iStart) * mBinSize;
-      long const iOffset = std::lround(offset / mBinSize * Float_t(parameters.mNResponseTables - 1));
-      if (iStart < 0) { // this should not happen
-        LOG(ERROR) << "FDDDigitizer: table lookup failure";
+
+    double hTime = t - getTOFCorrection(int(pmt / 4)); // account for TOF to detector
+    hTime += mIntRecord.timeNS;
+    o2::InteractionRecord irHit(hTime); // BC in which the hit appears (might be different from interaction BC for slow particles)
+
+    // nominal time of the BC to which the hit will be attributed
+    double bcTime = o2::InteractionRecord::bc2ns(irHit.bc, irHit.orbit);
+
+    int nCachedIR = 0;
+    for (int i = BCCacheMin; i < BCCacheMax + 1; i++) {
+      double tNS = hTime + o2::constants::lhc::LHCBunchSpacingNS * i;
+      cachedIR[nCachedIR].setFromNS(tNS);
+      if (tNS < 0 && cachedIR[nCachedIR] > irHit) {
+        continue; // don't go to negative BC/orbit (it will wrap)
       }
-      iStart = roundVc(std::max(long(0), iStart));
+      setBCCache(cachedIR[nCachedIR++]); // ensure existence of cached container
+    }
+    // if digit for this sector does not exist, create one otherwise add to it
+    Pulse(nPhE, hit.GetTrackID(), hTime, cachedIR, nCachedIR, pmt);
+
+  } //hit loop
+}
+
+//_____________________________________________________________________________
+void Digitizer::Pulse(int nPhE, int parID, double timeHit, std::array<o2::InteractionRecord, NBC2Cache> const& cachedIR, int nCachedIR, int channel)
+{
+  auto const roundVc = [&](int i) -> int {
+    return (i / Vc::float_v::Size) * Vc::float_v::Size;
+  };
 
-      Vc::float_v workVc;
-      Vc::float_v pmtVc;
-      Float_t const* q = mPMResponseTables[parameters.mNResponseTables / 2 + iOffset].data() + iStart;
-      Float_t const* qEnd = &mPMResponseTables[parameters.mNResponseTables / 2 + iOffset].back();
-      for (Int_t i = firstBin, iEnd = roundVc(lastBin); q < qEnd && i < iEnd; i += Vc::float_v::Size) {
+  double time0 = cachedIR[0].bc2ns(); // start time of the 1st cashed BC
+  float timeDiff = time0 - timeHit;
+  if (channel < 9)
+    timeDiff += parameters.mTimeDelayFDC;
+  else
+    timeDiff += parameters.mTimeDelayFDA;
+
+  //LOG(INFO) <<"Ch = "<<channel<<" NphE = " << nPhE <<" timeDiff "<<timeDiff;
+  Float_t charge = TMath::Qe() * parameters.mPmGain * mBinSize / mPmtTimeIntegral / mChargePerADC;
+
+  Bool_t added[nCachedIR];
```

Comment:
below you have lot of divisions in the loop by mBinSize, instead, define another `constexpr float BinSizeInv = 1./BinSize;`` and multiply by it.

Reply:
Done